xbmc/playlists/PlayListXSPF.cpp

```diff
@@ -0,0 +1,188 @@
+#define LOCATION_TAGNAME "location"
+#define PLAYLIST_TAGNAME "playlist"
+#define TITLE_TAGNAME "title"
+#define TRACK_TAGNAME "track"
+#define TRACKLIST_TAGNAME "trackList"
+
+#define missingElementMessage(E) "no " E " element"
+
+
+CPlayListXSPF::CPlayListXSPF(void) = default;
+
+CPlayListXSPF::~CPlayListXSPF(void) = default;
+
+
+bool CPlayListXSPF::Load(const std::string& strFileName)
+{
+  CXBMCTinyXML xmlDoc;
+  const char * errorStr = "";
+  int errorRow = 0;
+  int errorCol = 0;
+
+  const char * label = "";
+  const char * location = "";
+  int offset = 0;
+  std::string localpath;
+  CURL uri;
+
+  TiXmlElement* pPlaylist = nullptr;
+  TiXmlElement* pTracklist = nullptr;
+  TiXmlElement* pPlaylistTitle = nullptr;
+  TiXmlElement* pCurTrack = nullptr;
+
+  CLog::Log(LOGINFO, "ZZ Playlist=%s", strFileName);
+
+  if (!xmlDoc.LoadFile(strFileName))
+  {
+    goto error;
+  }
+
+  pPlaylist = xmlDoc.FirstChildElement(PLAYLIST_TAGNAME);
+  if (!pPlaylist)
+  {
+    errorStr = missingElementMessage(PLAYLIST_TAGNAME);
+    goto error;
+  }
+
+  pTracklist = pPlaylist->FirstChildElement(TRACKLIST_TAGNAME);
+  if (!pTracklist)
+  {
+    errorRow = pPlaylist->Row();
+    errorCol = pPlaylist->Column();
+    errorStr = missingElementMessage(TRACKLIST_TAGNAME);
+    goto error;
+  }
+
+  Clear();
+  URIUtils::GetParentPath(strFileName, m_strBasePath);
+
+  pPlaylistTitle = pPlaylist->FirstChildElement(TITLE_TAGNAME);
+  if (pPlaylistTitle)
+  {
+    m_strPlayListName = pPlaylistTitle->GetText();
+  }
+
+  pCurTrack = pTracklist->FirstChildElement(TRACK_TAGNAME);
+
+  while (pCurTrack)
+  {
+    label = "";
+    location = "";
+    localpath.clear();
+    uri.Reset();
+
+    TiXmlElement* pCurLocation = pCurTrack->FirstChildElement(LOCATION_TAGNAME);
+    if (pCurLocation)
+    {
+
+      TiXmlElement* pCurTitle = pCurTrack->FirstChildElement(TITLE_TAGNAME);
+      if (pCurTitle)
+      {
+        label = pCurTitle->GetText();
+      }
+
+      CFileItemPtr newItem(new CFileItem(label));
+
+      location = pCurLocation->GetText();
+      CLog::Log(LOGINFO, "Interpreting location=%s", location);
+
+      uri.Parse(location);
+
+      if (StringUtils::StartsWith(location, FILE_PROTOCOL_WITH_ROOT))
+      {
+#ifdef TARGET_WINDOWS
+        // omit leading / from "file:///C:/Users/..."
+	      offset = sizeof(FILE_PROTOCOL_WITH_ROOT) - 1;
+#else
+        // include leading / from "file:///home/..."
+	      offset = sizeof(FILE_PROTOCOL_WITH_ROOT) - 2;
+#endif
+
+        localpath = CURL::Decode(location + offset);
+      }
+      else if (uri.GetProtocol().empty())
+      {
+        localpath = URIUtils::AppendSlash(m_strBasePath) + CURL::Decode(location);
+      }
+
+      if (!localpath.empty())
+      {
+#ifdef TARGET_WINDOWS
+        StringUtils::Replace(localpath, "/", "\\");
+        localpath = URIUtils::CanonicalizePath(localpath, '\\');
+#else
+        localpath = URIUtils::CanonicalizePath(localpath, '/');
+#endif
```

Comment:
this looks hacky

Reply:
I'll take another pass at this. Is there a preferred format for storing paths in a platform-independent way? I found the `file://` protocol wasn't working as expected (was hoping that CFileItem->SetPath() on a `file://` would resolve correctly). Otherwise I'll see if I can keep to forward slashes with `URIUtils::CanonicalizePath(localpath, '/');`.

Reply:
It didn't get a lot better, but I'm hoping this is a little more clear. I've added comments describing why this pattern is being used.